jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -317,15 +319,133 @@ def learner_fn(runner_state: RunnerState) -> ExperimentOutput:
-            batched_update_step, runner_state, None, config["NUM_UPDATES"]
+        runner_state, (metric, loss_info) = jax.lax.scan(
+            batched_update_step, runner_state, None, config["NUM_UPDATES_PER_EVAL"]
         )
-        return {"runner_state": runner_state, "metrics": metric}
+        total_loss, (value_loss, loss_actor, entropy) = loss_info
+        return {
+            "runner_state": runner_state,
+            "metrics": metric,
+            "total_loss": total_loss,
+            "value_loss": value_loss,
+            "loss_actor": loss_actor,
+            "entropy": entropy,
+        }
 
     return learner_fn
 
 
-def learner_setup(env: Environment, config: Dict) -> Tuple[callable, RunnerState]:
+def get_evaluator_fn(env: Environment, apply_fn: callable, config: dict) -> callable:
+    """Get the evaluator function."""
+
+    def eval_one_episode(params, runner_state) -> Tuple:
+        """Evaluate one episode. It is vectorized over the number of evaluation episodes."""
+
+        def _env_step(runner_state: Tuple) -> Tuple:
+            """Step the environment."""
+            # PRNG keys.
+            rng, env_state, last_timestep, step_count_, return_ = runner_state
+
+            # Select action.
+            rng, _rng = jax.random.split(rng)
+            pi, _ = apply_fn(params, last_timestep.observation)
+
+            if config["EVALUATION_GREEDY"]:
+                action = pi.mode()
+            else:
+                action = pi.sample(seed=_rng)
+
+            # Step environment.
+            env_state, timestep = env.step(env_state, action)
+
+            # Log episode metrics.
+            return_ += timestep.reward
+            step_count_ += 1
+            runner_state = (rng, env_state, timestep, step_count_, return_)
+            return runner_state
+
+        def is_done(carry: Tuple) -> jnp.bool_:
+            """Check if the episode is done."""
+            timestep = carry[2]
+            return ~timestep.last()
+
+        rng, env_state, timestep = runner_state
+        return_ = jnp.array(0, float)
+        step_count_ = jnp.array(0, int)
+
+        final_runner = jax.lax.while_loop(
+            is_done,
+            _env_step,
+            (rng, env_state, timestep, step_count_, return_),
+        )
+
+        rng, env_state, timestep, step_count_, return_ = final_runner
+        eval_metrics = {
+            "episode_return": return_,
+            "episode_length": step_count_,
+        }
+        return eval_metrics
+
+    def evaluator_fn(
+        trained_params: FrozenDict, rng: chex.PRNGKey
+    ) -> Dict[str, Dict[str, chex.Array]]:
+        """Evaluator function."""
+
+        # Initialise environment states and timesteps.
+        n_devices = len(jax.devices())
+
+        eval_batch = config["NUM_EVAL_EPISODES"] // n_devices
+
+        rng, *env_rngs = jax.random.split(rng, eval_batch + 1)
+        env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
```

Comment:
in_axes=0 is default

```suggestion
        env_states, timesteps = jax.vmap(env.reset)(
```